src/main/java/io/druid/data/input/impl/JSONFlattenParser.java

```diff
@@ -0,0 +1,223 @@
+* under the License.
+*/
+
+package io.druid.data.input.impl;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.common.base.Charsets;
+import com.google.common.base.Function;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import com.metamx.common.parsers.ParseException;
+import com.metamx.common.parsers.Parser;
+import com.metamx.common.parsers.ParserUtils;
+
+import java.nio.charset.CharsetEncoder;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JSONFlattenParser implements Parser<String, Object>
+{
+  private static final Pattern arrayIndexRegex = Pattern.compile("(.*)(\\[\\d+\\])");
+  private final ObjectMapper objectMapper;
+  private ArrayList<String> fieldNames;
+  private final Set<String> exclude;
+  private final ParseFlattenSpec flattenSpec;
+  private Function<JsonNode, Object> valFn;
+
+  public static final Function<JsonNode, Object> flattenValueFunction = new Function<JsonNode, Object>()
+  {
+    @Override
+    public Object apply(JsonNode node)
+    {
+      if (node == null || node.isMissingNode() || node.isNull()) {
+        return null;
+      }
+      if (node.isIntegralNumber()) {
+        if (node.canConvertToLong()) {
+          return node.asLong();
+        } else {
+          return node.asDouble();
+        }
+      }
+      if (node.isFloatingPointNumber()) {
+        return node.asDouble();
+      }
+
+      final String s = node.asText();
+      final CharsetEncoder enc = Charsets.UTF_8.newEncoder();
+      if (s != null && !enc.canEncode(s)) {
+        // Some whacky characters are in this string (e.g. \uD900). These are problematic because they are decodeable
+        // by new String(...) but will not encode into the same character. This dance here will replace these
+        // characters with something more sane.
+        return new String(s.getBytes(Charsets.UTF_8), Charsets.UTF_8);
+      } else {
+        return s;
+      }
+    }
+  };
+
+  public JSONFlattenParser(
+      ObjectMapper objectMapper,
+      Iterable<String> fieldNames,
+      String timeColumn,
+      ParseFlattenSpec flattenSpec,
+      Function<JsonNode, Object> valFn
+  )
+  {
+    this(objectMapper, fieldNames, null, timeColumn, flattenSpec, valFn);
+  }
+
+  public JSONFlattenParser(
+      ObjectMapper objectMapper,
+      Iterable<String> fieldNames,
+      Iterable<String> exclude,
+      String timeColumn,
+      ParseFlattenSpec flattenSpec,
+      Function<JsonNode, Object> valFn
+  )
+  {
+    this.objectMapper = objectMapper;
+    if (fieldNames != null) {
+      setFieldNames(fieldNames);
+    }
+    this.fieldNames.add(timeColumn);
```

Comment:
This is going to throw a NullPointerException if `fieldNames` wasn't set.

I also don't think it's necessary- the purpose of `fieldNames` is to tell the parser that you only care about a certain specific set of fields. The parser constructor shouldn't be adding things to it.